### PR TITLE
Add Transpose+Gemm, Conv+BatchNormalization merging

### PR DESCRIPTION
**Pull request by take-cheeze**
- [x] Port [fuse_bn_into_conv.h](https://github.com/onnx/onnx/blob/abb1702ad38ee0671328116862821767cb389bd7/onnx/optimizer/passes/fuse_bn_into_conv.h), [fuse_transpose_into_gemm.h](https://github.com/onnx/onnx/blob/abb1702ad38ee0671328116862821767cb389bd7/onnx/optimizer/passes/fuse_transpose_into_gemm.h)
- [x] Support constant node input
- [x] Write tests
- [x] Disable this optimization on training, `node->outputs()->size() != 1`
- [x] Don't optimize detached node

### CODE REVIEW

**Comment by take-cheeze** [comment]
ChainerCI, test this please!

File: compiler/merge_test.cc [hunk 1]
```diff
@@ -70,5 +76,91 @@ TEST(MergeTest, PadConv) {
...
+    MergeOperations(&graph, false);
+    graph.DeleteDetached();
+    EXPECT_EQ(3, graph.nodes().size());
+    auto conv_checker = [](Node* nd) {
+        EXPECT_TRUE(nd->op_type() == Node::kConv || nd->op_type() == Node::kConstant);
+        return nd->op_type() == Node::kConv;
+    };
+    auto node_it = std::find_if(graph.nodes().begin(), graph.nodes().end(), conv_checker);
+    EXPECT_NE(graph.nodes().end(), node_it);
+    Node const& node = **node_it;
```
Comment (shinh) [inline compiler/merge_test.cc]: `const Node`

File: compiler/merge_test.cc [hunk 2]
```diff
@@ -70,5 +76,91 @@ TEST(MergeTest, PadConv) {
...
+    Graph graph("test");
+    Value* input = graph.AddInputValue("input", type);
+    Value* output = graph.AddOutputValue("output", type);
+
+    std::string trans_name;
+    {
+        GraphBuilder gb(&graph, "test", input);
+
+        // Transpose node
+        Value& trans = *gb.Op(Node::kTranspose, {input});
```
Comment (shinh) [inline compiler/merge_test.cc]: [optional] Please use pointers if you change https://github.com/pfnet-research/chainer-compiler/pull/334/files#diff-f3b721c5d8105e1df0219d67e6c87f99R183

File: compiler/merge_test.cc [hunk 3]
```diff
@@ -70,5 +76,91 @@ TEST(MergeTest, PadConv) {
...
+        trans_name = trans.name();
+        Node& trans_node = *trans.producer();
```
Comment (shinh) [inline compiler/merge_test.cc]: ditto

File: compiler/merge_test.cc [hunk 4]
```diff
@@ -70,5 +76,91 @@ TEST(MergeTest, PadConv) {
...
+    ASSERT_EQ(0, node.trans_b());
+    ASSERT_TRUE(std::none_of(node.inputs().begin(), node.inputs().end(), [trans_name](Value* v) { return v->name() == trans_name; }));
+    graph.CheckSanity("merged");
+}
+
 }  // namespace
 }  // namespace chainer_compiler
+
+// TODO(take-cheeze): EXPECT_ARRAY_* macros need to be used inside chainerx namespace.
+// These can move above when https://github.com/chainer/chainer/pull/7434 is merged.
```
Comment (shinh) [inline compiler/merge_test.cc]: Thanks!!!!!!!!

File: compiler/merge.cc [hunk 5]
```diff
@@ -119,20 +123,142 @@ bool MaybeMergePadConv(Graph* graph, Node* pad) {
...
+    chainerx::Array s = scale / chainerx::Sqrt(var + eps);
+    std::vector<chainerx::Array> new_w_data;
+    for (int64_t i = 0; i < w.shape()[0]; ++i) {
+        new_w_data.push_back(w.At({i}) * s.At({i}));
+    }
+    chainerx::Array new_w = chainerx::Stack(new_w_data);
+    bc = (bc - mean) * s + bn_bias;
+
+    GraphBuilder gb(graph, "MergeConvBN", bn->input(0));
+    Node& new_conv = *gb.MOp(Node::kConv, {conv->input(0), gb.Const(new_w), gb.Const(bc)}, bn->outputs());
```
[the remark this answers is not in the document]
Comment (take-cheeze) [reply compiler/merge.cc]: What I like about mutable reference it that:
- reference isn't re-assignable(or I don't need to specify another const after `*`)
- when lifetime analysis comes to C++, compiler should tell me more errors like rust

Though I think
- current code base is based on your style so it shouldn't be mixed for consistency
- lifetime analysis isn't here yet

now so I'll follow the current style.
Thanks for pointing out.

File: compiler/merge.cc [hunk 6]
```diff
@@ -119,20 +123,142 @@ bool MaybeMergePadConv(Graph* graph, Node* pad) {
     return true;
 }
 
+bool MaybeMergeConvBN(Graph* graph, Node* conv) {
+    if (conv->outputs().size() != 1) {
```
[the remark this answers is not in the document]
Comment (take-cheeze) [reply compiler/merge.cc]: Seems what I needed was detached node check:
https://github.com/pfnet-research/chainer-compiler/pull/334/commits/8f989d5650481a06c5b02eb6d03d6cd4f6ce2fa9
It conflicted with Pad+Conv merger.

File: compiler/merge.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (take-cheeze) [reply compiler/merge.cc]: Discussed this topic offline and we agreed to use mutable reference in local variable.
Though we still continue to use mutable pointer type in function argument.

**Comment by take-cheeze** [comment]
ChainerCI, test this please!
